src/lib/storage/value_segment.cpp

```diff
@@ -1,73 +1,94 @@
 
 template <typename T>
 AllTypeVariant ValueSegment<T>::operator[](const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  if (is_null(chunk_offset)) {
+    return NULL_VALUE;
+  }
+  return get(chunk_offset);
 }
 
 template <typename T>
 bool ValueSegment<T>::is_null(const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return is_nullable() && null_values()[chunk_offset];
 }
 
 template <typename T>
 T ValueSegment<T>::get(const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  Assert(!is_null(chunk_offset), "Tried to .get a NULL value from a ValueSegment.");
+  return values()[chunk_offset];
 }
 
 template <typename T>
 std::optional<T> ValueSegment<T>::get_typed_value(const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  if (is_null(chunk_offset)) {
+    return std::nullopt;
+  }
+  return get(chunk_offset);
 }
 
 template <typename T>
 void ValueSegment<T>::append(const AllTypeVariant& value) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  if (variant_is_null(value)) {
+    Assert(is_nullable(), "Tried to append NULL value into non-nullable ValueSegment.");
+    _values.emplace_back();
+    _null_values.emplace_back(true);
+  } else {
+    try {
+      _values.emplace_back(type_cast<T>(value));
+    } catch (...) {
+      try {
+        _values.emplace_back(boost::lexical_cast<T>(value));
+      } catch (...) {
+        Fail("Tried to append inconvertible value to ValueSegment.");
+      }
+    }
+
+    if (is_nullable()) {
+      _null_values.emplace_back(false);
+    }
+  }
 }
 
 template <typename T>
 ChunkOffset ValueSegment<T>::size() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return values().size();
 }
 
 template <typename T>
 const std::vector<T>& ValueSegment<T>::values() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _values;
 }
 
 template <typename T>
 bool ValueSegment<T>::is_nullable() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _is_nullable;
 }
 
 template <typename T>
 const std::vector<bool>& ValueSegment<T>::null_values() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  Assert(is_nullable(), "Tried to get null_values of non-nullable ValueSegment.");
+  return _null_values;
 }
 
 template <typename T>
 size_t ValueSegment<T>::estimate_memory_usage() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  // FIXME: From the tests, this seems to be what upstream wants. However, one could argue that the correct
+  //        implementation is
+  //
+  //            values().capacity() * sizeof(T) + null_values().capacity() / 8
```

Comment:
Yes, our group noticed this as well. According to the tests, you can actually at least use the capacity of the `values` vector in this computation, though.

Reply:
:shrug: we will leave it like this until maybe another test for another segment comes around to clarify